sdk/typescript/src/api/client.ts

```diff
@@ -0,0 +1,814 @@
+	SuiTransactionBlockResponse,
+	PaginatedEvents,
+	DevInspectResults,
+	PaginatedCoins,
+	SuiObjectResponse,
+	DelegatedStake,
+	CoinBalance,
+	CoinSupply,
+	Checkpoint,
+	CommitteeInfo,
+	DryRunTransactionBlockResponse,
+	SuiSystemStateSummary,
+	PaginatedObjectsResponse,
+	ValidatorsApy,
+	MoveCallMetrics,
+	ObjectRead,
+	ResolvedNameServiceNames,
+	ProtocolConfig,
+	EpochInfo,
+	EpochPage,
+	CheckpointPage,
+	DynamicFieldName,
+	DynamicFieldPage,
+	NetworkMetrics,
+	AddressMetrics,
+} from '../types/index.js';
+import {
+	isValidTransactionDigest,
+	isValidSuiAddress,
+	isValidSuiObjectId,
+	normalizeSuiAddress,
+	normalizeSuiObjectId,
+} from '../utils/sui-types.js';
+import { fromB58, toB64, toHEX } from '@mysten/bcs';
+import type { SerializedSignature } from '../cryptography/signature.js';
+import { TransactionBlock } from '../builder/index.js';
+import { SuiHTTPTransport } from './http-transport.js';
+import type { SuiNetwork, SuiTransport } from './http-transport.js';
+
+export interface PaginationArguments<Cursor> {
+	/** Optional paging cursor */
+	cursor?: Cursor;
+	/** Maximum item returned per page */
+	limit?: number | null;
+}
+
+export interface OrderArguments {
+	order?: Order | null;
+}
+
+/**
+ * Configuration options for the JsonRpcProvider. If the value of a field is not provided,
+ * value in `DEFAULT_OPTIONS` for that field will be used
+ */
+export type SuiApiClientOptions = NetworkOrTransport & {
+	/**
+	 * Cache timeout in seconds for the RPC API Version
+	 */
+	versionCacheTimeoutInSeconds?: number;
+};
+
+export type NetworkOrTransport =
+	| {
+			network: SuiNetwork;
+			transport?: never;
+	  }
+	| {
+			transport: SuiTransport;
+			network?: never;
+	  };
+
+const DEFAULT_OPTIONS = {
+	versionCacheTimeoutInSeconds: 600,
+} satisfies Partial<SuiApiClientOptions>;
+
+export class SuiAPIClient {
+	protected transport: SuiTransport;
+	private rpcApiVersion: string | undefined;
+	private cacheExpiry: number | undefined;
+	private versionCacheTimeoutInSeconds;
+	/**
+	 * Establish a connection to a Sui RPC endpoint
+	 *
+	 * @param options configuration options for the API Client
+	 */
+	constructor(options: SuiApiClientOptions) {
+		const opts = { ...DEFAULT_OPTIONS, ...options };
+		this.versionCacheTimeoutInSeconds = opts.versionCacheTimeoutInSeconds;
+		this.transport = opts.transport ?? new SuiHTTPTransport({ network: opts.network });
+	}
+
+	async getRpcApiVersion(): Promise<string | undefined> {
+		if (this.rpcApiVersion && this.cacheExpiry && this.cacheExpiry <= Date.now()) {
+			return this.rpcApiVersion;
+		}
+
+		try {
+			const resp = await this.transport.request<{ info: { version: string } }>({
+				method: 'rpc.discover',
+				params: [],
+			});
+			this.rpcApiVersion = resp.info.version;
+			this.cacheExpiry =
+				// Date.now() is in milliseconds, but the timeout is in seconds
+				Date.now() + (this.versionCacheTimeoutInSeconds ?? 0) * 1000;
+			return this.rpcApiVersion;
+		} catch (err) {
```

Comment:
I think you can remove the undefined return and just make this throw if it fails.